### PR TITLE
Skip iteration if pending pods are too new

### LINKED ISSUE #502: Wait another iteration if the oldest pending pod is just 1-2 seconds old
Autoscaling / NAP will give better results if it sees the whole picture - for example all pods created by the newly-created deployment, not just the initial fraction.

### PR DESCRIPTION
**Pull request by mwielgus**
Fixes #502

### CODE REVIEW

File: cluster-autoscaler/core/utils.go [hunk 1]
```diff
@@ -491,3 +491,13 @@ func UpdateClusterStateMetrics(csr *clusterstate.ClusterStateRegistry) {
 	readiness := csr.GetClusterReadiness()
 	metrics.UpdateNodesCount(readiness.Ready, readiness.Unready+readiness.LongNotStarted, readiness.NotStarted)
 }
+
+func getOldestCreateTime(pods []*apiv1.Pod) time.Time {
+	oldest := time.Now()
```
Comment (MaciekPytel) [inline cluster-autoscaler/core/utils.go]: I think in general we should use currentTime parameter of RunOnce, not time.Now(). Those timestamps may be different since we're already after the FilterOutSchedulable call, which can take noticeable amount of time. But we're still looking at cluster state as of 'currentTime' parameter.

**Comment by MaciekPytel** [comment]
I'm not entirely sold on this idea. I understand the intent - we can make more informed (ie. better) decisions when we see the whole picture and not just half of pods in deployment. On the other hand it will cost us in reaction time and we're adding yet another heuristic to already complex and hard to follow code. 

Do we know how often this would actually change anything? Is it common for CA to trigger in the middle of controller adding pods? Will that actually impact autoscaling decisions significantly (given that for a small number of pods pricing expander will mostly follow 'preferred cluster shape' part of equation?

If all of the above are true, than I fully agree with this idea. However, I'm worried that it will mostly apply to some edge case scenarios like adding large deployment of small pods with affinity in an already large cluster, at which point I'm not sure if it's worth the added complexity.

**Comment by mwielgus** [comment]
We cannot wait for "full picture". First of all it would be quite complex to tell if all deployments/statefull sets/jobs/whatever have created all of the required pods.
Second if the pods are created manually or via a broken controller then we may wait quite some time for the situation to stabilise, postponing scale up "forever".

This change is more targeted at NAP. With CA you most likely have just one or two node pools and the exact scale up doesn't matter. With nap, seeing just first 2-3 pods, may result in creating the wrong NodePool and sticking to it as we have a "2x" penalty on creating new node pool.

**Comment by MaciekPytel** [comment]
Sorry, I was apparently not clear - I don't think we should wait any more than this PR does. What I meant was quite opposite - is it worth waiting at all? 

Even for NAP, I'm not convinced seeing all pods (or just more pods) in deployment vs. seeing just one or two will impact node shape decisions all that much. It can change outcome in some complex scenarios (like my example with pod affinity above). However, I expect that in most cases the pricing difference is not that big and 'preferred cluster shape' part of formula will determine the outcome anyway.

My point is - I don't know what the real impact of such heuristic would be. I can find specific scenarios when I think it may help, but in normal use-cases I have no idea how often it will actually change anything (I'm not even sure how often it will actually trigger). Without any data it feels premature to add yet another heuristic.

**Comment by mwielgus** [comment]
AFAIR pod creation in deployment is throttled at 20 pod/second. With <10 pods current code will pick the consistent node pool or the existing one. With more than 20 pods chances are that the node pool will be suited to the new pods and not picked for consistency. Creating 20 pods takes 1 second. We run iteration every 10 seconds. 
So whenever we start scaling up for larger deployment, there is at least 10% probability that we are doing it very prematurely and we are breaking the chances of more suited cluster shape. 

If you wish I can give you more detailed calculations.

File: cluster-autoscaler/core/static_autoscaler.go [hunk 2]
```diff
@@ -244,6 +249,11 @@ func (a *StaticAutoscaler) RunOnce(currentTime time.Time) errors.AutoscalerError
 		glog.V(1).Info("No unschedulable pods")
 	} else if a.MaxNodesTotal > 0 && len(readyNodes) >= a.MaxNodesTotal {
 		glog.V(1).Info("Max total nodes in cluster reached")
+	} else if getOldestCreateTime(unschedulablePodsToHelp).Add(unschedulablePodTimeBuffer).After(currentTime) {
+		// The assumption here is that these pods have been created very recently and probably there
+		// is more pods to come. In theory we could check the newest pod time but then if pod were created
+		// slowly but at the pace of 1 every 2 seconds then no scale up would be triggered for long time.
+		glog.V(1).Info("Unschedulable pods are very new, waiting one iteration for more")
```
Comment (MaciekPytel) [inline cluster-autoscaler/core/static_autoscaler.go]: I think (not sure though) that without a return here we can actually do a scale-down in this loop. We don't consider unschedulable pods in scale-down logic (because previously we would only get there if len(unschedulablePodsToHelp) > 0 or no existing node group can fit any of the unschedulable pods).

**Comment by MaciekPytel** [comment]
I'm aware of the controller throttling and I agree with your numbers. I understand what this heuristic does and have some idea of when it will trigger and what it will do. And I agree it makes sense and it's actually quite simple.

On the other hand I'm reasonably confident that right now the set of people who knows what actually happens inside CA loop consists of exactly 2 people: you and me. And I'm starting to get lost among all of those heuristics too.

So I wonder if it's worth adding a heuristic that may potentially improve the shape of node CA / NAP picks if:
1. There is a deployment with specific enough requirements to need to pick for pod shape and not consistency.
1. This deployment gets created / scaled-up with >20 replicas delta.
1. No other pod is created in the same time period (as it could break 3s constraint).

Bottom line is - I'm not opposed to this heuristic as such. If your convinced it's the right thing to do I'm ok with it (after addressing my review comments). I'm just worried that the complexity of CA is slowly getting out of hand and it may be worth prioritizing readability over addressing every edge case we can think of before we actually see someone bumping into the problem.

File: cluster-autoscaler/core/utils.go (hunk 1 above)
Comment (mwielgus) [reply cluster-autoscaler/core/utils.go]: oldest is used only for safety, if the list of pods is empty, this value is unlikely to be used.

File: cluster-autoscaler/core/static_autoscaler.go (hunk 2 above)
Comment (mwielgus) [reply cluster-autoscaler/core/static_autoscaler.go]: Fixed.

File: cluster-autoscaler/core/utils.go (hunk 1 above)
Comment (MaciekPytel) [reply cluster-autoscaler/core/utils.go]: Right, I misunderstood how we use the timestamp. You're correct, it doesn't matter.

**Comment by MaciekPytel** [comment]
We've discussed offline and decided this heuristic is simple enough to have minimal impact on maintainability.

/lgtm
